texinfo fails around help2man (support `chmod()`)

Building the `texinfo` and `grub2-unsigned` packages fails around `help2man`. (Unless otherwise stated, I'm debugging `texinfo`'s compilation.)

The second build goes like

```
Making all in man
make[3]: Entering directory '/home/ubuntu/texi/texinfo-6.7.0.dfsg.2/man'
Updating man page makeinfo.1
cd ..//tp && /usr/bin/make makeinfo
make[4]: Entering directory '/home/ubuntu/texi/texinfo-6.7.0.dfsg.2/tp'
rm -f makeinfo
ln -s texi2any makeinfo
make[4]: Leaving directory '/home/ubuntu/texi/texinfo-6.7.0.dfsg.2/tp'
/bin/bash /home/ubuntu/texi/texinfo-6.7.0.dfsg.2/build-aux/missing help2man --name="translate Texinfo documents" -I ./makeinfo.h2m ..//tp/makeinfo -o makeinfo.1
help2man: can't get `--help' info from ..//tp/makeinfo
Try `--no-discard-stderr' if option outputs to stderr
make[3]: *** [Makefile:1416: makeinfo.1] Error 126
```

--- a/src/common/fbbcomm.def
+++ b/src/common/fbbcomm.def
@@ -352,7 +352,7 @@
       # file path
       (OPTIONAL, STRING, "path"),
       # mode
-      (OPTIONAL, "mode_t", "mode"),
+      (REQUIRED, "mode_t", "mode"),
       # flags for fchmodat(), AT_SYMLINK_NOFOLLOW for lchmod()
       (OPTIONAL, "int", "flags", "fbbcomm_debug_at_flags"),
       # error no., when ret = -1
@@ -361,9 +361,9 @@
 
     ("fchmod", [
       # file fd
-      (OPTIONAL, "int", "fd"),
+      (REQUIRED, "int", "fd"),
       # mode
-      (OPTIONAL, "mode_t", "mode"),
+      (REQUIRED, "mode_t", "mode"),
       # error no., when ret = -1
       (OPTIONAL, "int", "error_no"),
     ]),

--- a/src/firebuild/execed_process.cc
+++ b/src/firebuild/execed_process.cc
@@ -98,10 +98,11 @@ void ExecedProcess::initialize() {
   /* Propagate the opening of the executable and libraries upwards as
    * regular file open events. */
   if (parent_exec_point()) {
-    FileUsageUpdate exe_update = FileUsageUpdate::get_from_open_params(executable(), O_RDONLY, 0);
+    FileUsageUpdate exe_update =
+        FileUsageUpdate::get_from_open_params(executable(), O_RDONLY, 0, 0);
     parent_exec_point()->register_file_usage_update(executable(), exe_update);
     for (const auto& lib : libs_) {
-      FileUsageUpdate lib_update = FileUsageUpdate::get_from_open_params(lib, O_RDONLY, 0);
+      FileUsageUpdate lib_update = FileUsageUpdate::get_from_open_params(lib, O_RDONLY, 0, 0);
       parent_exec_point()->register_file_usage_update(lib, lib_update);
     }
   }

--- a/src/firebuild/execed_process_cacher.cc
+++ b/src/firebuild/execed_process_cacher.cc
@@ -326,24 +326,6 @@ static void add_file(std::vector<FBBSTORE_Builder_file>* files, const FileName* 
   }
 }
 
-static void add_file(std::vector<FBBSTORE_Builder_file>* files, const FileName* file_name,
-                     FileType type, const ssize_t content_size, const Hash *content_hash,
-                     const mode_t mode, const mode_t mode_mask) {
-  FBBSTORE_Builder_file& new_file = files->emplace_back();
-  new_file.set_path_with_length(file_name->c_str(), file_name->length());
-  new_file.set_type(type);
-  if (content_size >= 0) {
-    new_file.set_size(content_size);
-  }
-  if (content_hash) {
-    new_file.set_hash(content_hash->get());
-  }
-  if (mode_mask != 0) {
-    new_file.set_mode(mode);
-    new_file.set_mode_mask(mode_mask);
-  }
-}
-
 static const FBBSTORE_Builder* file_item_fn(int idx, const void *user_data) {
   const std::vector<FBBSTORE_Builder_file>* fbb_file_vector =
       reinterpret_cast<const std::vector<FBBSTORE_Builder_file> *>(user_data);
@@ -492,47 +474,83 @@ void ExecedProcessCacher::store(const ExecedProcess *proc) {
   for (const auto& pair : proc->file_usages()) {
     const auto filename = pair.first;
     const FileUsage* fu = pair.second;
+    /* fu contains information about the file's original contents or metadata, plus whether it's
+     * been modified. We need to query the current state of the file if anything has been modified. */
 
-    /* If the file's final contents matter, place it in the file cache,
-     * and also record it in pb's "outputs". This actually needs to
-     * compute the checksums now. */
-    if (fu->written()) {
-      int fd = open(filename->c_str(), O_RDONLY);
-      if (fd >= 0) {
-        struct stat64 st;
-        if (fstat64(fd, &st) == 0) {
-          if (S_ISREG(st.st_mode)) {
-            Hash new_hash;
-            /* TODO don't store and don't record if it was read with the same hash. */
+    if (!fu->written() && !fu->mode_changed()) {
+      /* Completely unchanged, nothing to do. */
+      continue;
+    }
+
+    FileInfo new_file_info(DONTKNOW);
+
+    struct stat64 st;
+    if (stat64(filename->c_str(), &st) == 0) {
+      /* We have something, let's see what it is. */
+      new_file_info.set_type(EXIST);
+
+      /* If the file's final contents matter, place it in the file cache,
+       * and also record it in pb's "outputs". This actually needs to
+       * compute the checksums now. */
+      if (fu->written()) {
+        if (S_ISREG(st.st_mode)) {
+          new_file_info.set_type(ISREG);
+          Hash new_hash;
+          /* TODO don't store and don't record if it was read with the same hash. */
+          int fd = open(filename->c_str(), O_RDONLY);
+          if (fd >= 0) {
             if (!hash_cache->store_and_get_hash(filename, &new_hash, fd, &st)) {
               /* unexpected error, now what? */
               FB_DEBUG(FB_DEBUG_CACHING,
                        "Could not store blob in cache, not writing shortcut info");
+              close(fd);
               return;
             }
-            // TODO(egmont) fail if setuid/setgid/sticky is set
-            int mode = st.st_mode & 07777;
-            add_file(&out_path_isreg, filename, ISREG, st.st_size, &new_hash, mode, 07777);
-          } else if (S_ISDIR(st.st_mode)) {
-            // TODO(egmont) fail if setuid/setgid/sticky is set
-            const int mode = st.st_mode & 07777;
-            add_file(&out_path_isdir, filename, ISDIR, -1, nullptr, mode, 07777);
-            out_path_isdir_filename_ptrs.insert(filename);
+            close(fd);
+            new_file_info.set_size(st.st_size);
+            new_file_info.set_hash(new_hash);
           } else {
-            // TODO(egmont) handle other types of entries
+            fb_perror("open");
+            new_file_info.set_type(NOTEXIST);
           }
+        } else if (S_ISDIR(st.st_mode)) {
+          new_file_info.set_type(ISDIR);
         } else {
-          fb_perror("fstat");
-          if (fu->initial_type() != NOTEXIST) {
-            out_path_notexist.push_back(filename->c_str());
-          }
+          // TODO(egmont) handle other types of entries
+          new_file_info.set_type(NOTEXIST);
         }
-        close(fd);
-      } else {
+      }
+
+      if (fu->mode_changed()) {
+        // TODO(egmont) fail if setuid/setgid/sticky is set
+        new_file_info.set_mode_bits(st.st_mode & 07777, 07777);
+      }
+    } else {
+      /* Stat failed, nothing at the new location. */
+      new_file_info.set_type(NOTEXIST);
+    }
+
+    switch (new_file_info.type()) {
+      case DONTKNOW:
+        /* This can happen if we figured out that the file didn't actually change. */
+        break;
+      case EXIST:
+      case ISREG:
+        // FIXME skip adding if the new state is the same as the old one
+        add_file(&out_path_isreg, filename, new_file_info);
+        break;
+      case ISDIR:
+        // FIXME skip adding if the new state is the same as the old one
+        add_file(&out_path_isdir, filename, new_file_info);
+        out_path_isdir_filename_ptrs.insert(filename);
+        break;
+      case NOTEXIST:
         if (fu->initial_type() != NOTEXIST) {
           out_path_notexist.push_back(filename->c_str());
         }
-      }
+        break;
+      default:
+        assert(0);
     }
   }
 
@@ -636,6 +654,20 @@ static FileInfo file_to_file_info(const FBBSTORE_Serialized_file *file) {
   }
   info.set_mode_bits(file->get_mode_with_fallback(0), file->get_mode_mask_with_fallback(0));
   return info;
+}
+
+/**
+ * Create a FileUsageUpdate object based on an FBB's File entry.
+ */
+static FileUsageUpdate file_to_file_usage_update(const FileName *filename,
+                                                 const FBBSTORE_Serialized_file *file) {
+  bool written = (file->get_type() == ISREG && file->has_size()) ||
+                 (file->get_type() == ISDIR);
+  bool mode_changed = file->has_mode();
+  /* "file" represents the file's _new_ state, so it's not suitable for the _initial_ state of the
+   * update. The initial state can be anything, i.e. DONTKNOW. */
+  FileUsageUpdate update(filename, DONTKNOW, written, mode_changed);
+  return update;
 }
 
 /**
@@ -824,8 +856,8 @@ static bool restore_dirs(
       }
     }
     if (proc->parent_exec_point()) {
-      proc->parent_exec_point()->register_file_usage_update(
-          path, FileUsageUpdate(path, DONTKNOW, true));
+      FileUsageUpdate update = file_to_file_usage_update(path, dir);
+      proc->parent_exec_point()->register_file_usage_update(path, update);
     }
   }
   return true;
@@ -870,8 +902,9 @@ static void remove_files_and_dirs(
       rmdir(path->c_str());
     }
     if (proc->parent_exec_point()) {
-      proc->parent_exec_point()->register_file_usage_update(
-          path, FileUsageUpdate(path, DONTKNOW, true));
+      // FIXME Register that it was an _empty_ directory
+      FileUsageUpdate update = FileUsageUpdate(path, ISDIR, true);
+      proc->parent_exec_point()->register_file_usage_update(path, update);
     }
   }
 }
@@ -920,19 +953,22 @@ bool ExecedProcessCacher::apply_shortcut(ExecedProcess *proc,
     const FBBSTORE_Serialized_file *file =
         reinterpret_cast<const FBBSTORE_Serialized_file *>(outputs->get_path_isreg_at(i));
     const auto path = FileName::Get(file->get_path(), file->get_path_len());
-    FB_DEBUG(FB_DEBUG_SHORTCUT,
-             "│   Fetching file from blobs cache: "
-             + d(path));
-    Hash hash(file->get_hash());
-    blob_cache->retrieve_file(hash, path);
+    if (file->get_type() == ISREG) {
+      FB_DEBUG(FB_DEBUG_SHORTCUT,
+               "│   Fetching file from blobs cache: "
+               + d(path));
+      assert(file->has_hash());
+      Hash hash(file->get_hash());
+      blob_cache->retrieve_file(hash, path);
+    }
     if (file->has_mode()) {
       /* Refuse to apply setuid, setgid, sticky bit. */
       // FIXME warn on them, even when we store them.
       chmod(path->c_str(), file->get_mode() & 0777);
     }
     if (proc->parent_exec_point()) {
-      proc->parent_exec_point()->register_file_usage_update(
-          path, FileUsageUpdate(path, DONTKNOW, true));
+      FileUsageUpdate update = file_to_file_usage_update(path, file);
+      proc->parent_exec_point()->register_file_usage_update(path, update);
     }
   }
 

--- a/src/firebuild/file_usage_update.cc
+++ b/src/firebuild/file_usage_update.cc
@@ -61,33 +61,6 @@ void FileUsageUpdate::type_computer_open_rdonly() const {
 }
 
 /**
- * If we saw a successful open(..., O_WRONLY|O_CREAT) (without O_TRUNC and O_EXCL; perhaps with
- * O_RDWR instead of O_RDONLY) then the following two cases can happen:
- * - Now the file is empty. We cannot tell if the file existed and was empty before, or did not exist.
- * - Now the file is non-empty. We know that the file existed before with the current contents.
- * This method performs the lazy on-demand check to see which of these two happened.
- */
-void FileUsageUpdate::type_computer_open_wronly_creat_notrunc_noexcl() const {
-  struct stat st;
-  if (stat(filename_->c_str(), &st) == -1) {
-    unknown_err_ = errno;
-    return;
-  }
-  if (st.st_size > 0) {
-    // FIXME handle if we see a directory. This cannot normally happen due to O_CREAT, but can if
-    // the file has just been replaced by a directory.
-    initial_state_.set_type(ISREG);
-    /* We got to know that this was a regular non-empty file. Delay hash computation until
-     * necessary. */
-    hash_computer_ = &FileUsageUpdate::hash_computer;
-  } else {
-    initial_state_.set_type(NOTEXIST_OR_ISREG);
-  }
-  initial_state_.set_size(st.st_size);
-  type_computer_ = nullptr;
-}
-
-/**
  * Get the file type, looking it up on demand if necessary.
  *
  * Due to the nature of lazy lookup, an unexpected error can occur, in which case false is returned.
@@ -149,8 +122,8 @@ bool FileUsageUpdate::get_initial_hash(Hash *hash_ptr) const {
  * can compute it on demand.
  */
 FileUsageUpdate FileUsageUpdate::get_from_open_params(const FileName *filename, int flags,
-                                                      int err) {
-  TRACK(FB_DEBUG_PROC, "flags=%d, err=%d", flags, err);
+                                                      mode_t mode_with_umask, int err) {
+  TRACK(FB_DEBUG_PROC, "flags=%d, mode_with_umask=0%03o, err=%d", flags, mode_with_umask, err);
 
   FileUsageUpdate update(filename);
 
@@ -168,8 +141,10 @@ FileUsageUpdate FileUsageUpdate::get_from_open_params(const FileName *filename, 
        */
       if ((flags & O_CREAT) && (flags & O_EXCL)) {
         /* C+F: If an exclusive new file was created, take a note that the file didn't exist
-         * previously, but its parent dir has to exist. */
+         * previously, that the permissions will have to be set on it, and that its parent dir has
+         * to exist. */
         update.set_initial_type(NOTEXIST);
+        update.mode_changed_ = true;
         update.parent_type_ = ISDIR;
       } else if (flags & O_TRUNC) {
         if (!(flags & O_CREAT)) {
@@ -180,7 +155,27 @@ FileUsageUpdate FileUsageUpdate::get_from_open_params(const FileName *filename, 
         } else {
           /* B: The old contents could have been any regular file, or even no such file (but not
            * e.g. a directory). Also, the parent directory has to exist. */
+          struct stat64 st;
+          if (stat64(filename->c_str(), &st) < 0) {
+            update.unknown_err_ = errno;
+            return update;
+          }
+          if (st.st_size > 0) {
+            /* We had O_TRUNC, so this is unexpected. */
+            update.unknown_err_ = EEXIST;
+            return update;
+          }
+          // FIXME handle if we see a directory. This cannot normally happen due to O_CREAT, but
+          // can if the file has just been replaced by a directory.
           update.set_initial_type(NOTEXIST_OR_ISREG);
+          if ((st.st_mode & 07777) != mode_with_umask) {
+            /* A mode mismatch implies that the file necessarily existed before. See #861. */
+            update.set_initial_type(ISREG);
+          } else {
+            /* The mode matches. The file may or may not have existed before, but in either case,
+             * it'll have the given permissions now. Pretend that they were set explicitly. #861. */
+            update.mode_changed_ = true;
+          }
           update.parent_type_ = ISDIR;
         }
       } else {
@@ -193,7 +188,30 @@ FileUsageUpdate FileUsageUpdate::get_from_open_params(const FileName *filename, 
           /* E: Another nasty combo. We can't distinguish a newly created empty file from a
            * previously empty one. If the file is non-empty, we need to store its hash. Also, the
            * parent directory has to exist. */
-          update.type_computer_ = &FileUsageUpdate::type_computer_open_wronly_creat_notrunc_noexcl;
+          struct stat64 st;
+          if (stat64(filename->c_str(), &st) < 0) {
+            update.unknown_err_ = errno;
+            return update;
+          }
+          if (st.st_size > 0) {
+            // FIXME handle if we see a directory. This cannot normally happen due to O_CREAT, but
+            // can if the file has just been replaced by a directory.
+            update.set_initial_type(ISREG);
+            /* We got to know that this was a regular non-empty file. Delay hash computation until
+             * necessary. */
+            update.hash_computer_ = &FileUsageUpdate::hash_computer;
+          } else {
+            update.set_initial_type(NOTEXIST_OR_ISREG);
+          }
+          update.set_initial_size(st.st_size);
+          if ((st.st_mode & 07777) != mode_with_umask) {
+            /* A mode mismatch implies that the file necessarily existed before. See #861. */
+            update.set_initial_type(ISREG);
+          } else {
+            /* The mode matches. The file may or may not have existed before, but in either case,
+             * it'll have the given permissions now. Pretend that they were set explicitly. #861. */
+            update.mode_changed_ = true;
+          }
           update.parent_type_ = ISDIR;
         }
       }
@@ -256,6 +274,7 @@ FileUsageUpdate FileUsageUpdate::get_from_mkdir_params(const FileName *filename,
     update.set_initial_type(NOTEXIST);
     update.parent_type_ = ISDIR;
     update.written_ = true;
+    update.mode_changed_ = true;
   } else {
     if (err == EEXIST) {
       /* The directory already exists. It may not be a directory, but in that case process inputs
@@ -311,13 +330,32 @@ FileUsageUpdate FileUsageUpdate::get_oldfile_usage_from_rename_params(const File
                                                                       const FileName *new_name,
                                                                       int error) {
   TRACK(FB_DEBUG_PROC, "err=%d", error);
-
   /* Read the file's hash from the new location, but update generation from the old one's name
    * to keep the generation number increasing. Otherwise it would be reset to 1, which is valid
    * for the newly created file (if the file did not exist before). */
   // TODO(rbalint) Error handling is way more complicated for rename than for open, fix that here.
-  FileUsageUpdate update = get_from_open_params(new_name, O_RDONLY, error);
+  FileUsageUpdate update = get_from_open_params(new_name, O_RDONLY, 0, error);
+  update.written_ = true;
+  update.mode_changed_ = true;
   update.generation_ = old_name->generation();
+
+  return update;
+}
+
+/**
+ * Based on the parameters and return value of a rename() or similar call, returns a FileUsageUpdate
+ * object that reflects how our usage of new file changed.
+ */
+FileUsageUpdate FileUsageUpdate::get_newfile_usage_from_rename_params(const FileName *new_name,
+                                                                      int error) {
+  TRACK(FB_DEBUG_PROC, "err=%d", error);
+
+  (void)error;  /* maybe unused */
+
+  /* The file at the new name now necessarily exists. It may or may not empty, it doesn't matter. We
+   * have to set mode_changed so that the mode will be restored when replaying from the cache.
+   * This does not match any of the A..F cases of get_from_open_params(). */
+  FileUsageUpdate update(new_name, DONTKNOW, true, true);
 
   return update;
 }

--- a/src/firebuild/file_usage_update.h
+++ b/src/firebuild/file_usage_update.h
@@ -24,12 +24,14 @@ class FileUsageUpdate {
       : initial_state_(type), filename_(filename), written_(written), mode_changed_(mode_changed),
         generation_(filename->generation()) {}
 
-  static FileUsageUpdate get_from_open_params(const FileName *filename, int flags, int err);
+  static FileUsageUpdate get_from_open_params(const FileName *filename, int flags,
+                                              mode_t mode_with_umask, int err);
   static FileUsageUpdate get_from_mkdir_params(const FileName *filename, int err);
   static FileUsageUpdate get_from_stat_params(const FileName *filename, mode_t mode, off_t size,
                                               int err);
   static FileUsageUpdate get_oldfile_usage_from_rename_params(const FileName* old_name,
                                                               const FileName* new_name, int err);
+  static FileUsageUpdate get_newfile_usage_from_rename_params(const FileName* new_name, int err);
 
   FileType parent_type() const {return parent_type_;}
   bool written() const {return written_;}

--- a/src/firebuild/firebuild.cc
+++ b/src/firebuild/firebuild.cc
@@ -889,8 +889,9 @@ void proc_ic_msg(const FBBCOMM_Serialized *fbbcomm_buf,
             const size_t path_len = action_open->get_path_len();
             int fd = action_open->get_fd();
             int flags = action_open->get_flags();
+            mode_t mode = action_open->get_mode();
             fork_child->handle_force_close(fd);
-            fork_child->handle_open(AT_FDCWD, path, path_len, flags, fd, 0);
+            fork_child->handle_open(AT_FDCWD, path, path_len, flags, mode, fd, 0);
             /* Revert the effect of "pre-opening" paths to be written in the posix_spawn message.*/
             if (is_write(flags)) {
               const firebuild::FileName* file_name = fork_child->get_absolute(AT_FDCWD, path,
@@ -1211,6 +1212,16 @@ void proc_ic_msg(const FBBCOMM_Serialized *fbbcomm_buf,
           reinterpret_cast<const FBBCOMM_Serialized_faccessat *>(fbbcomm_buf));
       break;
     }
+    case FBBCOMM_TAG_chmod: {
+      ::firebuild::ProcessFBBAdaptor::handle(proc,
+          reinterpret_cast<const FBBCOMM_Serialized_chmod *>(fbbcomm_buf));
+      break;
+    }
+    case FBBCOMM_TAG_fchmod: {
+      ::firebuild::ProcessFBBAdaptor::handle(proc,
+          reinterpret_cast<const FBBCOMM_Serialized_fchmod *>(fbbcomm_buf));
+      break;
+    }
     case FBBCOMM_TAG_memfd_create: {
       ::firebuild::ProcessFBBAdaptor::handle(proc,
           reinterpret_cast<const FBBCOMM_Serialized_memfd_create *>(fbbcomm_buf));
@@ -1279,11 +1290,9 @@ void proc_ic_msg(const FBBCOMM_Serialized *fbbcomm_buf,
       proc->exec_point()->disable_shortcutting_bubble_up("clone() is not supported");
       break;
     }
-    case FBBCOMM_TAG_chmod:
     case FBBCOMM_TAG_chown:
     case FBBCOMM_TAG_fb_debug:
     case FBBCOMM_TAG_fb_error:
-    case FBBCOMM_TAG_fchmod:
     case FBBCOMM_TAG_fchown:
     case FBBCOMM_TAG_fpathconf:
     case FBBCOMM_TAG_ftruncate:

--- a/src/firebuild/process.cc
+++ b/src/firebuild/process.cc
@@ -176,12 +176,12 @@ int Process::handle_pre_open(const int dirfd, const char * const ar_name, const 
 }
 
 int Process::handle_open(const int dirfd, const char * const ar_name, const size_t ar_len,
-                         const int flags, const int fd, const int error,
+                         const int flags, const mode_t mode, const int fd, const int error,
                          int fd_conn, const int ack_num, const bool pre_open_sent) {
   TRACKX(FB_DEBUG_PROC, 1, 1, Process, this,
-         "dirfd=%d, ar_name=%s, flags=%d, pre_open_sent=%d, fd=%d, error=%d, fd_conn=%s, "
-         "ack_num=%d",
-         dirfd, D(ar_name), flags, fd, pre_open_sent, error, D_FD(fd_conn), ack_num);
+         "dirfd=%d, ar_name=%s, flags=%d, mode=0%03o, pre_open_sent=%d, fd=%d, error=%d, "
+         "fd_conn=%s, ack_num=%d",
+         dirfd, D(ar_name), flags, mode, fd, pre_open_sent, error, D_FD(fd_conn), ack_num);
 
   const FileName* name = get_absolute(dirfd, ar_name, ar_len);
   if (!name) {
@@ -207,7 +207,8 @@ int Process::handle_open(const int dirfd, const char * const ar_name, const size
     name->close_for_writing();
   }
 
-  FileUsageUpdate update = FileUsageUpdate::get_from_open_params(name, flags, error);
+  FileUsageUpdate update = FileUsageUpdate::get_from_open_params(name, flags,
+                                                                 mode & 07777 & ~umask(), error);
   if (!exec_point()->register_file_usage_update(name, update)) {
     exec_point()->disable_shortcutting_bubble_up("Could not register the opening of a file", *name);
     if (ack_num != 0) {
@@ -239,7 +240,7 @@ int Process::handle_freopen(const char * const ar_name, const size_t ar_len,
     handle_close(oldfd, 0);
 
     /* Register the opening of the new file, no matter if succeeded or failed. */
-    return handle_open(AT_FDCWD, ar_name, ar_len, flags, fd, error, fd_conn,
+    return handle_open(AT_FDCWD, ar_name, ar_len, flags, 0666, fd, error, fd_conn,
                        ack_num, pre_open_sent);
   } else {
     /* Without a name pre_open should not have been sent. */
@@ -264,7 +265,7 @@ int Process::handle_freopen(const char * const ar_name, const size_t ar_len,
 
       /* Register the reopening, no matter if succeeded or failed. */
       return handle_open(AT_FDCWD, filename->c_str(), filename->length(),
-                         flags, fd, error, fd_conn, ack_num, pre_open_sent);
+                         flags, 0666, fd, error, fd_conn, ack_num, pre_open_sent);
     }
   }
 }
@@ -534,6 +535,80 @@ int Process::handle_faccessat(const int dirfd, const char * const ar_name, const
   if (!exec_point()->register_file_usage_update(name, update)) {
     exec_point()->disable_shortcutting_bubble_up(
         "Could not register the faccessat of a file", *name);
+    return -1;
+  }
+
+  return 0;
+}
+
+int Process::handle_chmod(const int dirfd, const char * const ar_name, const size_t ar_len,
+                          const mode_t mode, const int flags, const int error) {
+  TRACKX(FB_DEBUG_PROC, 1, 1, Process, this,
+         "dirfd=%d, ar_name=%s, mode=%d, flags=%d, error=%d",
+         dirfd, D(ar_name), mode, flags, error);
+
+  (void)mode;  /* No need to remember what we chmod to, we'll stat at the end. */
+
+  /* Chmod doesn't actually support AT_EMPTY_PATH, I guess it's a bug in the Linux kernel. */
+  const FileName* name =
+      ((flags & AT_EMPTY_PATH) && (ar_name[0] == '\0')) ? get_fd_filename(dirfd)
+      : get_absolute(dirfd, ar_name, ar_len);
+
+  if (!name) {
+    // FIXME don't disable shortcutting if chmod() failed due to the invalid dirfd
+    exec_point()->disable_shortcutting_bubble_up(
+        "Invalid dirfd or filename passed to chmod() variant");
+    return -1;
+  }
+
+  if (error) {
+    exec_point()->disable_shortcutting_bubble_up("Cannot register a failed chmod", *name);
+    return -1;
+  }
+
+  FileUsageUpdate update(name, EXIST, false, true);
+  if (!exec_point()->register_file_usage_update(name, update)) {
+    exec_point()->disable_shortcutting_bubble_up("Could not register a chmod", *name);
+    return -1;
+  }
+
+  return 0;
+}
+
+int Process::handle_fchmod(const int fd, const mode_t mode, const int error) {
+  TRACKX(FB_DEBUG_PROC, 1, 1, Process, this,
+         "fd=%d, mode=%d, error=%d", fd, mode, error);
+
+  (void)mode;  /* No need to remember what we chmod to, we'll stat at the end. */
+
+  const FileFD *file_fd = get_fd(fd);
+  if (!file_fd) {
+    if (error == 0) {
+      exec_point()->disable_shortcutting_bubble_up(
+          "Process fchmod()ed an unknown fd successfully, "
+          "which means interception missed at least one open()", fd);
+      return -1;
+    } else {
+      /* Invalid fd passed to fchmod(), or something like that. */
+      return 0;
+    }
+  }
+
+  const FileName* name = file_fd->filename();
+  if (!name) {
+    exec_point()->disable_shortcutting_bubble_up(
+        "Invalid fd or filename passed to fchmod()");
+    return -1;
+  }
+
+  if (error) {
+    exec_point()->disable_shortcutting_bubble_up("Cannot register a failed fchmod", *name);
+    return -1;
+  }
+
+  FileUsageUpdate update(name, EXIST, false, true);
+  if (!exec_point()->register_file_usage_update(name, update)) {
+    exec_point()->disable_shortcutting_bubble_up("Could not register a fchmod", *name);
     return -1;
   }
 
@@ -834,7 +909,7 @@ int Process::handle_rename(const int olddirfd, const char * const old_ar_name,
   /* Register the opening for writing at the new location */
   // TODO(rbalint) fix error handling, it is way more complicated for rename than for open.
   FileUsageUpdate update_new =
-      FileUsageUpdate::get_from_open_params(new_name, O_CREAT|O_WRONLY|O_TRUNC, error);
+      FileUsageUpdate::get_newfile_usage_from_rename_params(new_name, error);
   if (!exec_point()->register_file_usage_update(new_name, update_new)) {
     exec_point()->disable_shortcutting_bubble_up(
         "Could not register the renaming (to)", *new_name);

--- a/src/firebuild/process.h
+++ b/src/firebuild/process.h
@@ -232,6 +232,7 @@ class Process {
    * @param ar_name relative or absolute file name
    * @param ar_len length of ar_name
    * @param flags flags of open()
+   * @param mode mode (create permissions) of open()
    * @param fd the return value, or -1 if file was dlopen()ed successfully
    * @param error error code of open()
    * @param fd_conn fd to send ACK on when needed
@@ -239,7 +240,7 @@ class Process {
    * @param pre_open_sent interceptor already sent pre_open for this open
    */
   int handle_open(const int dirfd, const char * const ar_name, const size_t ar_len, const int flags,
-                  const int fd, const int error = 0, int fd_conn = -1,
+                  const mode_t mode, const int fd, const int error = 0, int fd_conn = -1,
                   int ack_num = 0, const bool pre_open_sent = false);
 
   /**
@@ -331,6 +332,11 @@ class Process {
    */
   int handle_faccessat(const int dirfd, const char * const name, const size_t name_len,
                        const int mode, const int flags, const int error = 0);
+
+  int handle_chmod(const int dirfd, const char * const name, const size_t name_len,
+                   const mode_t mode, const int flags, const int error = 0);
+
+  int handle_fchmod(const int fd, const mode_t mode, const int error = 0);
 
   /**
    * Handle memfd_create in the monitored process

--- a/src/firebuild/process_fbb_adaptor.cc
+++ b/src/firebuild/process_fbb_adaptor.cc
@@ -20,7 +20,8 @@ int ProcessFBBAdaptor::handle(Process *proc, const FBBCOMM_Serialized_open *msg,
   int error = msg->get_error_no_with_fallback(0);
   int ret = msg->get_ret_with_fallback(-1);
   return proc->handle_open(dirfd, msg->get_file(), msg->get_file_len(), msg->get_flags(),
-                           ret, error, fd_conn, ack_num, msg->get_pre_open_sent());
+                           msg->get_mode_with_fallback(0), ret, error, fd_conn, ack_num,
+                           msg->get_pre_open_sent());
 }
 
 int ProcessFBBAdaptor::handle(Process *proc, const FBBCOMM_Serialized_freopen *msg, int fd_conn,
@@ -37,7 +38,7 @@ int ProcessFBBAdaptor::handle(Process *proc, const FBBCOMM_Serialized_dlopen *ms
   if (!msg->has_error_string() && msg->has_absolute_filename()) {
     return proc->handle_open(AT_FDCWD,
                              msg->get_absolute_filename(), msg->get_absolute_filename_len(),
-                             O_RDONLY, -1, 0, fd_conn, ack_num);
+                             O_RDONLY, 0, -1, 0, fd_conn, ack_num);
   } else {
     std::string filename = msg->has_absolute_filename() ? msg->get_absolute_filename() : "NULL";
     proc->exec_point()->disable_shortcutting_bubble_up("Process failed to dlopen() ", filename);
@@ -102,6 +103,22 @@ int ProcessFBBAdaptor::handle(Process *proc, const FBBCOMM_Serialized_faccessat 
   const int error = msg->get_error_no_with_fallback(0);
   return proc->handle_faccessat(dirfd, msg->get_pathname(), msg->get_pathname_len(),
                                 mode, flags, error);
+}
+
+int ProcessFBBAdaptor::handle(Process *proc, const FBBCOMM_Serialized_chmod *msg) {
+  const int dirfd = msg->get_dirfd_with_fallback(AT_FDCWD);
+  const mode_t mode = msg->get_mode();
+  const int flags = msg->get_flags_with_fallback(0);
+  const int error = msg->get_error_no_with_fallback(0);
+  return proc->handle_chmod(dirfd, msg->get_path(), msg->get_path_len(),
+                            mode, flags, error);
+}
+
+int ProcessFBBAdaptor::handle(Process *proc, const FBBCOMM_Serialized_fchmod *msg) {
+  const int fd = msg->get_fd();
+  const int mode = msg->get_mode();
+  const int error = msg->get_error_no_with_fallback(0);
+  return proc->handle_fchmod(fd, mode, error);
 }
 
 int ProcessFBBAdaptor::handle(Process *proc, const FBBCOMM_Serialized_memfd_create *msg) {

--- a/src/firebuild/process_fbb_adaptor.h
+++ b/src/firebuild/process_fbb_adaptor.h
@@ -31,6 +31,8 @@ class ProcessFBBAdaptor {
   static int handle(Process *proc, const FBBCOMM_Serialized_fstat *msg);
   static int handle(Process *proc, const FBBCOMM_Serialized_stat *msg);
   static int handle(Process *proc, const FBBCOMM_Serialized_faccessat *msg);
+  static int handle(Process *proc, const FBBCOMM_Serialized_chmod *msg);
+  static int handle(Process *proc, const FBBCOMM_Serialized_fchmod *msg);
   static int handle(Process *proc, const FBBCOMM_Serialized_memfd_create *msg);
   static int handle(Process *proc, const FBBCOMM_Serialized_timerfd_create *msg);
   static int handle(Process *proc, const FBBCOMM_Serialized_epoll_create *msg);

